cmd/rocsp-tool/main.go

```diff
@@ -118,31 +154,285 @@ func main2() error {
-	ctx := context.Background()
-	respBytes, err := ioutil.ReadFile(respFile)
+func helpExit() {
+	fmt.Fprintf(os.Stderr, "Usage: %s [store|copy-from-db] --config path/to/config.json\n", os.Args[0])
+	fmt.Fprintln(os.Stderr, "  store -- for each command line arg, read that filename as an OCSP response and store it in Redis")
+	fmt.Fprintln(os.Stderr, "  load-from-db -- scan the database for all OCSP entries for unexpired certificates, and store in Redis")
+	fmt.Fprintln(os.Stderr)
+	flag.PrintDefaults()
+	os.Exit(1)
+}
+
+func configureOCSPGenerator(tlsConf cmd.TLSConfig, grpcConf cmd.GRPCClientConfig, clk clock.Clock, stats prometheus.Registerer) (capb.OCSPGeneratorClient, error) {
+	tlsConfig, err := tlsConf.Load()
 	if err != nil {
-		return fmt.Errorf("reading response file %q: %w", respFile, err)
+		return nil, fmt.Errorf("loading TLS config: %w", err)
 	}
+	clientMetrics := bgrpc.NewClientMetrics(stats)
+	caConn, err := bgrpc.ClientSetup(&grpcConf, tlsConfig, clientMetrics, clk)
+	cmd.FailOnError(err, "Failed to load credentials and create gRPC connection to CA")
+	return capb.NewOCSPGeneratorClient(caConn), nil
+}
+
+func configureDb(dbConfig *cmd.DBConfig) (*sql.DB, error) {
+	if dbConfig == nil {
+		return nil, nil
+	}
+	dsn, err := dbConfig.URL()
+	if err != nil {
+		return nil, fmt.Errorf("loading DB URL: %w", err)
+	}
+
+	conf, err := mysql.ParseDSN(dsn)
+	if err != nil {
+		return nil, fmt.Errorf("while parsing DSN from 'DBConnectFile': %s", err)
+	}
+
+	if len(conf.Params) == 0 {
+		conf.Params = make(map[string]string)
+	}
+	conf.Params["tx_isolation"] = "'READ-UNCOMMITTED'"
+	conf.Params["interpolateParams"] = "true"
+	conf.Params["parseTime"] = "true"
+
+	db, err := sql.Open("mysql", conf.FormatDSN())
+	if err != nil {
+		return nil, fmt.Errorf("couldn't setup database client: %s", err)
+	}
+
+	db.SetMaxOpenConns(dbConfig.MaxOpenConns)
+	db.SetMaxIdleConns(dbConfig.MaxIdleConns)
+	db.SetConnMaxLifetime(dbConfig.ConnMaxLifetime.Duration)
+	db.SetConnMaxIdleTime(dbConfig.ConnMaxIdleTime.Duration)
+	return db, nil
+}
+
+// setDefault sets the target to a default value, if it is zero.
+func setDefault(target *int, def int) {
+	if *target == 0 {
+		*target = def
+	}
+}
+
+type client struct {
+	issuers       []ShortIDIssuer
+	redis         *rocsp.WritingClient
+	db            *sql.DB // optional
+	ocspGenerator capb.OCSPGeneratorClient
+	clk           clock.Clock
+}
+
+func (cl *client) loadFromDB(ctx context.Context, speed ProcessingSpeed) error {
+	// To scan the DB efficiently, we want to select only currently-valid certificates. There's a
+	// handy expires index, but for selecting a large set of rows, using the primary key will be
+	// more efficient. So first we find a good id to start with, then scan from there. Note: since
+	// AUTO_INCREMENT can skip around a bit, we add padding to ensure we get all currently-valid
+	// certificates.
```

Comment:
Same, honestly. I don't have a strong upper bound here. What I'd _prefer_ to do is maintain some sort of "low water mark" of the lowest ID that has made it fully through the system to storage. I couldn't think of a particularly great way to do that, but I'm interested in ideas!

Reply:
Alright, updated so we now emit the "lowest inflight ID", and also have the query use `>=` so that would be a reasonable thing to restart from.

Note: This does not _currently_ implement a "start from id X" flag. I'd like to do that as a followup, along with cleaning up stdout so it's not so spammy.

Reply:
Can we reference https://github.com/letsencrypt/boulder/issues/5783 as a `TODO(#5783)` here?